Operations review

dlindhol opened the pull request:
This should be good enough for our first review of the `Operations` implementation. It ended up being more about having `SampledFunctions` applying the pure functions that `Operations` provide.  I named the branch "algebra" because I was hoping to define a `FunctionalAlgebra` trait or such that could be implemented by Datasets and Adapters.... That became a distraction so I returned to the `Operation` centric perspective so we could get something working. I expect it will evolve as we learn more.

The key driver was the `RddFunction` in `latis-hylatis` and continuing to support the Hysics use case.
Previously, Hylatis used a `SparkAdapter` to find a cached RDD and apply operations to it. The `algebra` branch of `latis-hylatis` now encapsulates the RDD in an `RddFunction` which implements `SampledFunction`. A Dataset can be "cached" to Spark yielding a new Dataset backed by an `RddFunction`. Operations on the Dataset then ultimately delegate to the Spark RDD. 

I ended up doing more implementation than I intended in order to keep Hylatis working. I cut a lot of corners assuming this limited use case. Lots of fodder for new tickets to keep us busy. I smell monoids, functors, and monads!

lindholc commented on README.md:
```diff
@@ -1,5 +1,52 @@
+* **fdml** - (Functional Datamodel Markup Language) An xml file format to describe the contents of a *Dataset*.  
+See the FDML Schema section below.
+* **Function** - A *Function* is a *DataType* consisting of exactly two *DataType* objects, a domain and a range. Function is a member of the DataType algebraic data type.
+* **functional** **data** **model** - Datsets consist of ordered sequences of samples that are decoupled from metadata.  Operations that implement the functional algebra can be performed on these samples to create new datasets.
```
I believe `**functional data model**` will work just fine here.

There's probably also more that can be said about the FDM.

sapols commented:
I'm not sure that I agree with _everything_ Chris said. I think I do like the idea of making a separate document for the glossary; I can see how it could clutter up the README. I like the design decisions in the README. But I don't think we need to move anything from the glossary into Scaladocs. Sure, adding stuff to Scaladocs is a good idea, but I think we can do that and still keep the entries in the glossary. I like the idea of having a "one-stop-shop" glossary for our terms better than having to go combing through documentation to find something.

sapols commented on README.md:
```diff
@@ -1,5 +1,52 @@
+# Summary
+
 *LaTiS version 3* is a rewrite of the *Functional Data Model* and *Algebra* based on lessons learned from version 2, and with a bit more inspiration from Functional Programming. It will incubate as *latis3 version 0.x* until we are ready to finalize the API and release it as *latis version 3.x*.
```
A little nitpicky, but we should be consistent with how we capitalize "LaTiS". For example, you wrote "Latis3 version 0.x" and I think that should be "LaTiS3 version 0.x".

sapols commented on README.md:
```diff
@@ -1,5 +1,52 @@
+Items that are capitalized are Scala class names.
+
+* **arity** - The number of un-flattened *DataTypes* in the domain of a function.  
+For example, a function with a domain of (x, y) and wavelength has an arity of 2.
```
Maybe "...with a domain of (x, y) **and a range of** wavelength..."?

sapols replied:
Ahh! I misread.

sapols commented on README.md:
```diff
@@ -1,5 +1,52 @@
+For example, a function with a domain of (x, y) and wavelength has an arity of 2.
+* **Data** - A recursive algebraic data type consisting of *ScalarData*, *TupleData*, and *FunctionData*. Unlike the *DataType*, which describes the model, the *Data* objects contain actual values representing measurements.  Since *TupleData* and *FunctionData* contain multiple *Data* objects they can be represented as trees just like DataType objects.
+* **Dataset** - A *Dataset* instance consists of three things:
+  1. *MetaData* for the dataset itself. A list of key/value pairs.
```
Maybe use a colon? "*MetaData* for the dataset itself: A list of key/value pairs."

sapols commented on README.md:
```diff
@@ -1,5 +1,52 @@
+* **DataType** - A recursive algebraic data type consisting of *Scalars*, *Tuples*, and *Functions*.  The *DataType* is purely in the realm of the model.  Since *Tuples* and Functions themselves contain multiple *DataType* objects they can be nested with arbitrary complexity, meaning a *DataType* can be represented as a tree.  
+Also, every *DataType* must contain a *MetaData* object.
+* **dimensionality** - The number of flattened *DataTypes* in the domain of a function.  
+For example, a function with a domain of (x, y) and wavelength has a dimensionality of 3.
```
Maybe "...with a domain of (x, y) **and a range of** wavelength..."?

sapols commented on README.md:
```diff
@@ -1,5 +1,52 @@
+  * Con: Seems arbitrary to require both a single *Data* item followed by a list of additional *Data* items.
+  * Pro: Ensures that at least one Data item is in the *TupleData*.
+* *Sample* objects are not implemented as a tuple of domain and range, as a *Function* is, but are instead implemented as an integer representing dimensionality followed by the domain and range combined into a single list:  
+  * Con: The dimensionality number feels  arbitrary.  simply breaking a *Sample* into domain and range would be more elegant.
```
Accidental double space between "feels" and "arbitrary".

pfisterp commented on README.md:
```diff
@@ -1,5 +1,52 @@
+* **Function** - A *Function* is a *DataType* consisting of exactly two *DataType* objects, a domain and a range. Function is a member of the DataType algebraic data type.
+* **functional data model** - Datsets consist of ordered sequences of samples that are decoupled from metadata.  Operations that implement the functional algebra can be performed on these samples to create new datasets.
+* **Metadata** - An instance of *Metadata* is simply a list of key/value pairs to describe a *DataType* or *Dataset*.  Every *Scalar* object is required to contain a *MetaData* object, but *MetaData* objects may be empty for *Tuple* and *Function* objects.
+* **range** - The dependant variables of a function.
```
Yeah, we don't celebrate Independance Day do we?

sapols replied:
Maybe we should though... 💃

sapols commented on README.md:
```diff
@@ -42,7 +42,7 @@ An example is a 2D point consisting of the x,y *Tuple*.  The scalars x and y are
   * Pro: Ensures that at least one Data item is in the *TupleData*.
 * *Sample* objects are not implemented as a tuple of domain and range, as a *Function* is, but are instead implemented as an integer representing dimensionality followed by the domain and range combined into a single list:  
-  * Con: The dimensionality number feels  arbitrary.  simply breaking a *Sample* into domain and range would be more elegant.
+  * Con: The dimensionality number feels arbitrary.  simply breaking a *Sample* into domain and range would be more elegant.
```
There's still a double space before the word "simply", which still isn't capitalized. :)

sapols commented:
I went ahead and approved this PR. I'll let the discussion about the glossary get resolved under Chris's "requested changes".

sapols commented on README.md:
```diff
@@ -1,52 +1,13 @@
-  * Pro: The model contains all the information needed to represent nested tuples.
-
-
+see glossary [here](docs/glossary.md)
```
Good catch! That totally slipped passed me, but yes, "here" or "see more" links are notorious accessibility issues.

dlindhol committed:
Add the basic implementation for Operations

This includes significant modifications to SampleFunction and its
implementations that enable the functional algebra.

sapols commented on core/src/main/scala/latis/output/Writer.scala:
```diff
@@ -16,6 +16,10 @@ class Writer(out: OutputStream) {
   //TODO: encode to bytes or T as opposed to String
   
+  /*
+   * TODO: have SampledFunction toString produce the "asc" output?
```
I like this idea.

sapols commented on core/src/main/scala/latis/ops/Operation.scala:
```diff
@@ -1,12 +1,75 @@
+   * Provide an indication that this operation was applied 
+   * for the provenance (history) metadata.
+   */
+  def provenance: String = this.toString //TODO: add time, version
```
Cool.

dlindhol commented:
I added a Substitution operation that might be worthy of some special review:
https://github.com/latis-data/latis3/pull/3/files#diff-a007cbff78b787750b3ae4db7d756336

Feel free to comment on anything else in this PR. I plan to merge soon and make more tickets to clean up the cruft.

lindholc commented on core/src/main/scala/latis/ops/Substitution.scala:
```diff
@@ -0,0 +1,76 @@
+  
+  
+  def applyToData(ds1: Dataset, ds2: Dataset): SampledFunction = {
+    // Get the dt2 domain variable id
```
`s/dt2/ds2/`?

pfisterp reviewed (commented):
I like the short functions that only do one thing.

pfisterp commented on core/src/main/scala/latis/data/MemoizedFunction.scala:
```diff
@@ -1,7 +1,128 @@
+    //TODO: implicit Interpolation strategy
+    //TODO: take advantage of ordering, find should at least short-circuit
+    samples find {
+      case Sample(d, _) => DomainOrdering.equiv(d, v) //Note, can't use "=="
```
Variable names d and v should be renamed to something more descriptive.

dlindhol replied:
When doing a pattern match, I tend to use little names if they are immediately used. When matching on `Sample` that is often "d" (domain) and "r" (range). That could be a reasonable convention. "v" (value) however, is definitely pushing its luck. In this case it comes from a method argument which could be the more verbose "value".

pfisterp commented on core/src/main/scala/latis/data/Sample.scala:
```diff
@@ -1,27 +1,53 @@
+      else ??? //TODO: error
+  }
   
+//  def zipPositionWithValue: Seq[(SamplePosition, Any)] = {
```
Just a reminder, this probably shouldn't be here.

dlindhol replied:
Indeed, we should avoid commented out code. In this case, I felt like I needed to be able to zip the sample position with it's value to support easy access of values during an operation. I then found another way for the specific case but thought this might still be needed. I suspect we can make that call sooner than later.

lindholc commented on core/src/main/scala/latis/data/ArrayFunction1D.scala:
```diff
@@ -0,0 +1,22 @@
+    case _ => ??? //new RuntimeException("Failed to evaluate ArrayFunction1D")
+  }
+  
+  def samples: Seq[Sample] = {
```
Another way you could write this that came to mind:

```scala
Seq.tabulate(array.length) { i => Sample(DomainData(i), array(i)) }
```

dlindhol replied:
I do like that better. It also applies to our `ArrayFunction2D`. Or should that be `Array2DFunction`? I've been leaning towards `FooBarFunction` for these subclasses (e.g. RddFunction). Too bad we can't say `2DArrayFunction`.

lindholc commented on core/src/main/scala/latis/data/IndexedFunction1D.scala:
```diff
@@ -14,16 +14,17 @@ case class IndexedFunction1D(as: Array[Any], vs: Array[Any]) extends MemoizedFun
-    val ss = (as zip vs).map(p => (DomainData(p._1), RangeData(p._2)))
-    Stream.emits(ss)
+  def samples: Seq[Sample] = {
+    (as zip vs).map(p => Sample(DomainData(p._1), RangeData(p._2)))
```
I think you could write this instead if you wanted to avoid `_1` and `_2`:

```scala
(as zip vs).map { case (a, v) => Sample(DomainData(a), RangeData(v)) }
```

dlindhol replied:
Yep. I seem to get mixed results when I try to sneak a `case` in without a `match`. I'll get used to it one of these times.
When breaking the `()` off of `map` into `{}`, I prefer to drop the `.` also.
```(as zip vs) map { case (a, v) => Sample(DomainData(a), RangeData(v)) }```

lindholc commented on core/src/main/scala/latis/data/MemoizedFunction.scala:
```diff
@@ -1,7 +1,128 @@
+    samples find {
+      case Sample(d, _) => DomainOrdering.equiv(d, v) //Note, can't use "=="
+      case _ => false
+    } match {
```
I've seen the pattern of matching on an `Option`, applying `f` if there is something there or else returning `x`, in several places.

Another way to do this is

```scala
.map(f(_)).getOrElse(x)
```

which is the same as

```scala
.fold(x)(f(_))
```

So I think this could be written

```scala
.fold(SampledFunction.empty)(SampledFunction(_))
```

dlindhol replied:
I'm going to sleep on this one. I'd certainly like to make better use of folds but I worry about readability. Once I master them, then I will feel elitist enough to expect everyone else to understand them. :-)

dlindhol replied:
Actually, I do see pattern matching on Option as a code smell. I recently got used to mapping a function over an Option so I'm beginning to better appreciate its monadic nature. Folding to optionally build up from an empty SampledFunction sounds defensible.

lindholc commented on core/src/main/scala/latis/data/MemoizedFunction.scala:
```diff
@@ -1,7 +1,128 @@
+   */
+  def unapply(sf: SampledFunction): Option[Seq[Sample]] = sf match {
+    case mf: MemoizedFunction => 
+      if (mf.isEmpty) None 
```
I also see this pattern a lot: a predicate that determines whether we get `Some` or `Nothing`, or something similar.

With cats we can do something like this:

```scala
Alternative[Option].guard(! mf.isEmpty).as(mf.samples)
```

It's not as compelling as I hoped it would be.

dlindhol replied:
Interesting. See note above on readability. :-)
